### PR TITLE
introduce Cqlproxy test

### PR DESCRIPTION
**Pull request by amoskong**

### CODE REVIEW

File: cqlproxy_test.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    """
+
+    def _build(self):
+        # git clone request permission
+        proxy.remoter.run('git clone https://github.com/scylladb/cqlproxy')
+        proxy.remoter.run('cd cqlproxy/dist; make release/rpm')
+        result = proxy.remoter.run('find ./release/rpm/|grep rpm$')
+        return result.stdout.strip()
+
+    def setup(self):
```
Comment (bentsi) [inline cqlproxy_test.py]: set**U**p ('U' upper case)

File: cqlproxy_test.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+        self.proxy = cassandra_seed
+
+        #rpm_path = _build()
+        rpm_path = 'https://s3.amazonaws.com/scylla-qa-team/cqlproxy-latest.x86_64.rpm'
+        proxy.remoter.run('sudo yum install -y %s' % rpm_path)
+
+        proxy.remoter.run("echo 'config_file: /etc/cassandra/conf/cassandra.yaml\nreplica_address: {}:9042\nlisten_address: {}:3042\n' > /tmp/cqlproxy.yaml ".format(
+                          self.scylla_seed.private_ip_address, self.proxy.private_ip_address))
+        proxy.remoter.run('sudo mv /tmp/cqlproxy.yaml /etc/cqlproxy/cqlproxy.yaml')
+        proxy.remoter.run('sudo systemctl restart cqlproxy', ignore_status=True)
```
Comment (bentsi) [inline cqlproxy_test.py]: why are you ignoring the status?

File: cqlproxy_test.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    def test_stop_wait_start(self):
+        base_cmd = self.params.get('stress_cmd')
+        # run a workload
+        stress_queue = self.run_stress_thread(stress_cmd=base_cmd, stress_num=2)
+
+        node = self.db_cluster.nodes[-1]
+        node.remoter.run('sudo systemctl stop scylla-server.service')
+        node.wait_db_down()
+        time.sleep(30)
```
Comment (bentsi) [inline cqlproxy_test.py]: why sleeping here?

Comment (amoskong) [reply cqlproxy_test.py]: Workload will write more data in this stage.

File: cqlproxy_test.py (hunk 2 above)
Comment (amoskong) [reply cqlproxy_test.py]: It's for debug, I will remove it. Thanks

File: tests/aws-cqlproxy.yaml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+# scylla-manager configuration
+# if running on aws and use_mgmt is true, the monitor image should not contain scylla
+use_mgmt: true
+mgmt_port: 10090
+scylla_repo_m: 'http://repositories.scylladb.com/scylla/repo/f4a2920f80c4bf178217c2553ad65ad7/centos/scylladb-2017.1.repo'
+scylla_mgmt_repo: 'http://downloads.scylladb.com.s3.amazonaws.com/manager/rpm/unstable/centos/branch-1.0/latest/scylla-manager.repo'
+
+backends: !mux
+    gce: !mux
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: If it's not supported on GCE, please remove it from backends.

File: tests/aws-cqlproxy.yaml [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+n_loaders: 1
+n_monitor_nodes: 1
+nemesis_class_name: 'ChaosMonkey'
+nemesis_interval: 5
+user_prefix: 'cqlproxy-test-amos-tmp'
+failure_post_behavior: keep
+space_node_threshold: 644245094
+ip_ssh_connections: 'public'
+experimental: 'true'
+append_conf: 'enable_sstable_data_integrity_check: true'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: remove this, because it causes stalls.

File: tests/aws-cqlproxy.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
+test_duration: 60
+stress_cmd: "cassandra-stress mixed cl=QUORUM duration=60m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port native=3042,jmx=6868 -mode cql3 native -rate threads=20 -pop seq=1..100000000 -log interval=5"
+n_db_nodes: 3
+n_loaders: 1
+n_monitor_nodes: 1
+nemesis_class_name: 'ChaosMonkey'
+nemesis_interval: 5
+user_prefix: 'cqlproxy-test-amos-tmp'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Change the user prefix to: cqlproxy-VERSION

File: tests/aws-cqlproxy.yaml [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+failure_post_behavior: keep
+space_node_threshold: 644245094
+ip_ssh_connections: 'public'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Why public?

File: tests/aws-cqlproxy.yaml [hunk 8]
```diff
@@ -0,0 +1,77 @@
+test_duration: 60
+stress_cmd: "cassandra-stress mixed cl=QUORUM duration=60m -schema 'replication(factor=3) compaction(strategy=SizeTieredCompactionStrategy)' -port native=3042,jmx=6868 -mode cql3 native -rate threads=20 -pop seq=1..100000000 -log interval=5"
+n_db_nodes: 3
+n_loaders: 1
+n_monitor_nodes: 1
+nemesis_class_name: 'ChaosMonkey'
+nemesis_interval: 5
+user_prefix: 'cqlproxy-test-amos-tmp'
+failure_post_behavior: keep
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: why keep?

File: tests/aws-cqlproxy.yaml [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+# aws instance_provision: on_demand|spot_fleet|spot_low_price|spot_duration
+instance_provision: 'on_demand'
+
+# scylla-manager configuration
+# if running on aws and use_mgmt is true, the monitor image should not contain scylla
+use_mgmt: true
+mgmt_port: 10090
+scylla_repo_m: 'http://repositories.scylladb.com/scylla/repo/f4a2920f80c4bf178217c2553ad65ad7/centos/scylladb-2017.1.repo'
+scylla_mgmt_repo: 'http://downloads.scylladb.com.s3.amazonaws.com/manager/rpm/unstable/centos/branch-1.0/latest/scylla-manager.repo'
+
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Do you really need mgmt?

File: tests/aws-cqlproxy.yaml [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        gce_root_disk_type_monitor: 'pd-standard'
+        gce_root_disk_size_monitor: 50
+        gce_n_local_ssd_disk_monitor: 0
+        scylla_repo: 'https://s3.amazonaws.com/downloads.scylladb.com/rpm/unstable/centos/branch-1.7/17/scylla.repo'
+        us_east_1:
+          gce_datacenter: 'us-east1-b'
+    aws: !mux
+        cluster_backend: 'aws'
+        instance_type_loader: 'c4.4xlarge'
+        instance_type_monitor: 'c4.2xlarge'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Monitor node shouldn't be that big (t2.small is enough here).

File: tests/aws-cqlproxy.yaml [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        us_east_1:
+            region_name: 'us-east-1'
+            security_group_ids: 'sg-c5e1f7a0'
+            subnet_id: 'subnet-d934e980'
+            ami_id_db_scylla: 'ami-a2bc65df'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: replace with AMI_ID

File: tests/aws-cqlproxy.yaml [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+            ami_id_monitor: 'ami-29035f52'
+            ami_monitor_user: 'centos'
+
+databases: !mux
+    cassandra:
+        db_type: cassandra
+        instance_type_db: 'm3.large'
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i2.4xlarge'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Move ami_id_db_ to here

Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Change to i3.2xlarge

File: tests/aws-cqlproxy.yaml [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+            ami_db_cassandra_user: 'ubuntu'
+            #ami_id_db_cassandra: 'ami-3eff0028'
+            #ami_db_cassandra_user: 'centos'
+            ami_id_monitor: 'ami-29035f52'
+            ami_monitor_user: 'centos'
+
+databases: !mux
+    cassandra:
+        db_type: cassandra
+        instance_type_db: 'm3.large'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: move ami_db_cassandra_ to here

Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Is it big enough?

File: tests/aws-cqlproxy.yaml [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    scylla:
+        db_type: scylla
+        instance_type_db: 'i2.4xlarge'
+    mixed:
+        db_type: mixed
+        instance_type_db: 'i2.4xlarge'
```
Comment (roydahan) [inline tests/aws-cqlproxy.yaml]: Is type_db needed also here?

File: cqlproxy_test.py [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+        proxy.remoter.run('sudo systemctl status cqlproxy')
+        proxy.remoter.run('sudo systemctl status cqlproxy_shadow')
+
+    def test_basic_cs(self):
+        # -port native=3042 -node proxy.ip
+        base_cmd = self.params.get('stress_cmd')
+        base_cmd += ' -port native=9042 -node %s' % self.cassandra_seed.private_ip_address
+        stress_queue = self.run_stress_thread(stress_cmd=base_cmd, stress_num=2)
+        self.verify_stress_thread(stress_queue)
+
```
Comment (roydahan) [inline cqlproxy_test.py]: Are you running this directly to cassandra without the proxy?
Is the proxy set in this stage?

File: cqlproxy_test.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+        base_cmd = self.params.get('stress_cmd')
+        base_cmd += ' -port native=3042 -node %s' % self.proxy.private_ip_address
+        stress_queue = self.run_stress_thread(stress_cmd=base_cmd, stress_num=2)
+        self.verify_stress_thread(stress_queue)
+
```
Comment (roydahan) [inline cqlproxy_test.py]: We need verification that make sure the data in on both clusters...

File: cqlproxy_test.py [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+    def test_stop_wait_start(self):
```
Comment (roydahan) [inline cqlproxy_test.py]: These tests aren't needed for now.
let's first have the basic test working.

File: cqlproxy_test.py [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+        node = self.db_cluster.nodes[-1]
+        node.remoter.run('sudo systemctl stop scylla-server.service')
+        node.wait_db_down()
+        time.sleep(30)
+        node.remoter.run('sudo systemctl start scylla-server.service')
+        node.wait_db_up()
+
+        self.verify_stress_thread(stress_queue)
+
+    def test_restart(self):
```
Comment (roydahan) [inline cqlproxy_test.py]: same (not needed now)

File: cqlproxy_test.py [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+        proxy.remoter.run('sudo yum install -y %s' % rpm_path)
+
+        proxy.remoter.run("echo 'config_file: /etc/cassandra/conf/cassandra.yaml\nreplica_address: {}:9042\nlisten_address: {}:3042\n' > /tmp/cqlproxy.yaml ".format(
+                          self.scylla_seed.private_ip_address, self.proxy.private_ip_address))
+        proxy.remoter.run('sudo mv /tmp/cqlproxy.yaml /etc/cqlproxy/cqlproxy.yaml')
+        proxy.remoter.run('sudo systemctl restart cqlproxy')
+        proxy.remoter.run('sudo systemctl status cqlproxy')
+        proxy.remoter.run('sudo systemctl status cqlproxy_shadow')
+
+    def test_basic_cs(self):
```
Comment (roydahan) [inline cqlproxy_test.py]: Test should look as follow:
1. Run load on cassandra only.
2. (Migrate the existing data to scylla) - can skip at the moment.
3. Set the proxy.
4. make sure data is started to be written in Scylla.
5. stop load and compare data in both clusters.

**Comment by roydahan** [comment]
@amoskong, please fix and send new PR

**Pull request closed without merging**